# Hometask 3: Text classification on BBC
*Author: Marina Talyanskaya*

# Data preparation.

As I failed to properly download already prepared data, I decied to take the raw one and to write it into csv-file with two columns: news texts and classes, using some tips from the web.

In [174]:
# Writing CSV-file with news texts and classes
import os
import pandas as pd

data_folder = "C://For Python/bbc"
folders = ["business","entertainment","politics","sport","tech"]

os.chdir(data_folder)

x = []
y = []

for i in folders:
    files = os.listdir(i)
    for text_file in files:
        file_path = i + "/" +text_file
        print("reading file:", file_path)
        with open(file_path) as f:
            data = f.readlines()
        data = ' '.join(data)
        x.append(data)
        y.append(i)
   
data = {'news': x, 'type': y}       
df = pd.DataFrame(data)
print ('writing csv flie ...')
df.to_csv('../dataset.csv', index=False)

reading file: business/001.txt
reading file: business/002.txt
reading file: business/003.txt
reading file: business/004.txt
reading file: business/005.txt
reading file: business/006.txt
reading file: business/007.txt
reading file: business/008.txt
reading file: business/009.txt
reading file: business/010.txt
reading file: business/011.txt
reading file: business/012.txt
reading file: business/013.txt
reading file: business/014.txt
reading file: business/015.txt
reading file: business/016.txt
reading file: business/017.txt
reading file: business/018.txt
reading file: business/019.txt
reading file: business/020.txt
reading file: business/021.txt
reading file: business/022.txt
reading file: business/023.txt
reading file: business/024.txt
reading file: business/025.txt
reading file: business/026.txt
reading file: business/027.txt
reading file: business/028.txt
reading file: business/029.txt
reading file: business/030.txt
reading file: business/031.txt
reading file: business/032.txt
reading 

reading file: business/271.txt
reading file: business/272.txt
reading file: business/273.txt
reading file: business/274.txt
reading file: business/275.txt
reading file: business/276.txt
reading file: business/277.txt
reading file: business/278.txt
reading file: business/279.txt
reading file: business/280.txt
reading file: business/281.txt
reading file: business/282.txt
reading file: business/283.txt
reading file: business/284.txt
reading file: business/285.txt
reading file: business/286.txt
reading file: business/287.txt
reading file: business/288.txt
reading file: business/289.txt
reading file: business/290.txt
reading file: business/291.txt
reading file: business/292.txt
reading file: business/293.txt
reading file: business/294.txt
reading file: business/295.txt
reading file: business/296.txt
reading file: business/297.txt
reading file: business/298.txt
reading file: business/299.txt
reading file: business/300.txt
reading file: business/301.txt
reading file: business/302.txt
reading 

reading file: entertainment/028.txt
reading file: entertainment/029.txt
reading file: entertainment/030.txt
reading file: entertainment/031.txt
reading file: entertainment/032.txt
reading file: entertainment/033.txt
reading file: entertainment/034.txt
reading file: entertainment/035.txt
reading file: entertainment/036.txt
reading file: entertainment/037.txt
reading file: entertainment/038.txt
reading file: entertainment/039.txt
reading file: entertainment/040.txt
reading file: entertainment/041.txt
reading file: entertainment/042.txt
reading file: entertainment/043.txt
reading file: entertainment/044.txt
reading file: entertainment/045.txt
reading file: entertainment/046.txt
reading file: entertainment/047.txt
reading file: entertainment/048.txt
reading file: entertainment/049.txt
reading file: entertainment/050.txt
reading file: entertainment/051.txt
reading file: entertainment/052.txt
reading file: entertainment/053.txt
reading file: entertainment/054.txt
reading file: entertainment/

reading file: entertainment/259.txt
reading file: entertainment/260.txt
reading file: entertainment/261.txt
reading file: entertainment/262.txt
reading file: entertainment/263.txt
reading file: entertainment/264.txt
reading file: entertainment/265.txt
reading file: entertainment/266.txt
reading file: entertainment/267.txt
reading file: entertainment/268.txt
reading file: entertainment/269.txt
reading file: entertainment/270.txt
reading file: entertainment/271.txt
reading file: entertainment/272.txt
reading file: entertainment/273.txt
reading file: entertainment/274.txt
reading file: entertainment/275.txt
reading file: entertainment/276.txt
reading file: entertainment/277.txt
reading file: entertainment/278.txt
reading file: entertainment/279.txt
reading file: entertainment/280.txt
reading file: entertainment/281.txt
reading file: entertainment/282.txt
reading file: entertainment/283.txt
reading file: entertainment/284.txt
reading file: entertainment/285.txt
reading file: entertainment/

reading file: politics/118.txt
reading file: politics/119.txt
reading file: politics/120.txt
reading file: politics/121.txt
reading file: politics/122.txt
reading file: politics/123.txt
reading file: politics/124.txt
reading file: politics/125.txt
reading file: politics/126.txt
reading file: politics/127.txt
reading file: politics/128.txt
reading file: politics/129.txt
reading file: politics/130.txt
reading file: politics/131.txt
reading file: politics/132.txt
reading file: politics/133.txt
reading file: politics/134.txt
reading file: politics/135.txt
reading file: politics/136.txt
reading file: politics/137.txt
reading file: politics/138.txt
reading file: politics/139.txt
reading file: politics/140.txt
reading file: politics/141.txt
reading file: politics/142.txt
reading file: politics/143.txt
reading file: politics/144.txt
reading file: politics/145.txt
reading file: politics/146.txt
reading file: politics/147.txt
reading file: politics/148.txt
reading file: politics/149.txt
reading 

reading file: politics/384.txt
reading file: politics/385.txt
reading file: politics/386.txt
reading file: politics/387.txt
reading file: politics/388.txt
reading file: politics/389.txt
reading file: politics/390.txt
reading file: politics/391.txt
reading file: politics/392.txt
reading file: politics/393.txt
reading file: politics/394.txt
reading file: politics/395.txt
reading file: politics/396.txt
reading file: politics/397.txt
reading file: politics/398.txt
reading file: politics/399.txt
reading file: politics/400.txt
reading file: politics/401.txt
reading file: politics/402.txt
reading file: politics/403.txt
reading file: politics/404.txt
reading file: politics/405.txt
reading file: politics/406.txt
reading file: politics/407.txt
reading file: politics/408.txt
reading file: politics/409.txt
reading file: politics/410.txt
reading file: politics/411.txt
reading file: politics/412.txt
reading file: politics/413.txt
reading file: politics/414.txt
reading file: politics/415.txt
reading 

reading file: sport/261.txt
reading file: sport/262.txt
reading file: sport/263.txt
reading file: sport/264.txt
reading file: sport/265.txt
reading file: sport/266.txt
reading file: sport/267.txt
reading file: sport/268.txt
reading file: sport/269.txt
reading file: sport/270.txt
reading file: sport/271.txt
reading file: sport/272.txt
reading file: sport/273.txt
reading file: sport/274.txt
reading file: sport/275.txt
reading file: sport/276.txt
reading file: sport/277.txt
reading file: sport/278.txt
reading file: sport/279.txt
reading file: sport/280.txt
reading file: sport/281.txt
reading file: sport/282.txt
reading file: sport/283.txt
reading file: sport/284.txt
reading file: sport/285.txt
reading file: sport/286.txt
reading file: sport/287.txt
reading file: sport/288.txt
reading file: sport/289.txt
reading file: sport/290.txt
reading file: sport/291.txt
reading file: sport/292.txt
reading file: sport/293.txt
reading file: sport/294.txt
reading file: sport/295.txt
reading file: sport/

reading file: tech/048.txt
reading file: tech/049.txt
reading file: tech/050.txt
reading file: tech/051.txt
reading file: tech/052.txt
reading file: tech/053.txt
reading file: tech/054.txt
reading file: tech/055.txt
reading file: tech/056.txt
reading file: tech/057.txt
reading file: tech/058.txt
reading file: tech/059.txt
reading file: tech/060.txt
reading file: tech/061.txt
reading file: tech/062.txt
reading file: tech/063.txt
reading file: tech/064.txt
reading file: tech/065.txt
reading file: tech/066.txt
reading file: tech/067.txt
reading file: tech/068.txt
reading file: tech/069.txt
reading file: tech/070.txt
reading file: tech/071.txt
reading file: tech/072.txt
reading file: tech/073.txt
reading file: tech/074.txt
reading file: tech/075.txt
reading file: tech/076.txt
reading file: tech/077.txt
reading file: tech/078.txt
reading file: tech/079.txt
reading file: tech/080.txt
reading file: tech/081.txt
reading file: tech/082.txt
reading file: tech/083.txt
reading file: tech/084.txt
r

reading file: tech/351.txt
reading file: tech/352.txt
reading file: tech/353.txt
reading file: tech/354.txt
reading file: tech/355.txt
reading file: tech/356.txt
reading file: tech/357.txt
reading file: tech/358.txt
reading file: tech/359.txt
reading file: tech/360.txt
reading file: tech/361.txt
reading file: tech/362.txt
reading file: tech/363.txt
reading file: tech/364.txt
reading file: tech/365.txt
reading file: tech/366.txt
reading file: tech/367.txt
reading file: tech/368.txt
reading file: tech/369.txt
reading file: tech/370.txt
reading file: tech/371.txt
reading file: tech/372.txt
reading file: tech/373.txt
reading file: tech/374.txt
reading file: tech/375.txt
reading file: tech/376.txt
reading file: tech/377.txt
reading file: tech/378.txt
reading file: tech/379.txt
reading file: tech/380.txt
reading file: tech/381.txt
reading file: tech/382.txt
reading file: tech/383.txt
reading file: tech/384.txt
reading file: tech/385.txt
reading file: tech/386.txt
reading file: tech/387.txt
r

In [175]:
data = pd.read_csv('C://For Python/dataset.csv',encoding='utf8')
data.head()

news      type
0  Ad sales boost Time Warner profit\n \n Quarter...  business
1  Dollar gains on Greenspan speech\n \n The doll...  business
2  Yukos unit buyer faces loan claim\n \n The own...  business
3  High fuel prices hit BA's profits\n \n British...  business
4  Pernod takeover talk lifts Domecq\n \n Shares ...  business

In [176]:
#Adding new column with numerical values for each class and converting it to 'dict'object type
data['category_id'] = data['type'].factorize()[0]

In [177]:
from io import StringIO
category_id_df = data[['type', 'category_id']].sort_values('category_id')
category_to_id = dict(category_id_df.values)
id_to_category = dict(category_id_df[['category_id', 'type']].values)

In [178]:
data.head()

news      type  category_id
0  Ad sales boost Time Warner profit\n \n Quarter...  business            0
1  Dollar gains on Greenspan speech\n \n The doll...  business            0
2  Yukos unit buyer faces loan claim\n \n The own...  business            0
3  High fuel prices hit BA's profits\n \n British...  business            0
4  Pernod takeover talk lifts Domecq\n \n Shares ...  business            0

Let's see if our classes are balanced or not. As far as we can see on the graph, classes are more or less balanced, each of topic is represented by 400-500 news.

In [314]:
import matplotlib.pyplot as plt
fig = plt.figure(figsize=(8,6))
df.groupby('type').news.count().plot.bar(ylim=0)
plt.show()

KeyError: (0.0, 265.5, 'business', 'black', 'center_baseline', 'right', 6953815889054266475, 'vertical', None, 72.0, <weakref at 0x000001AD1ACCD2C8; dead>, 1.2)

<Figure size 576x432 with 1 Axes>

In [180]:
#Converting text to vectorized representation. 
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(sublinear_tf=True, min_df=5, norm='l2', encoding='latin-1', ngram_range=(1, 2), stop_words='english')

features = tfidf.fit_transform(df.news).toarray()
labels = data.category_id
features.shape

(2225, 14454)

I use TFiDF method instead of bag of words as for some topics where could be some common words and collactions (f.g. 'data analysis' for tech and business or 'economy' for politics and business). Thus, TFiDF approach takes this issue into account and helps to define 'the most specific' words and collacations.
Each of 2225 news texts is represented by 14454 features, representing the TFiDF score for different words and collacations.

Let's apply Chi-2 methode to our news texts to define the most correlated terms for each of topics.
Actually, it computes chi-squared stats between each non-negative feature and class.
All found words and collacations seem to be relevant for the topic.

In [181]:
from sklearn.feature_selection import chi2
import numpy as np

N = 2
for Product, category_id in sorted(category_to_id.items()):
  features_chi2 = chi2(features, labels == category_id)
  indices = np.argsort(features_chi2[0])
  feature_names = np.array(tfidf.get_feature_names())[indices]
  unigrams = [v for v in feature_names if len(v.split(' ')) == 1]
  bigrams = [v for v in feature_names if len(v.split(' ')) == 2]
  print("# '{}':".format(Product))
  print("  . Most correlated words:\n       . {}".format('\n       . '.join(unigrams[-N:])))
  print("  . Most correlated collacatios:\n       . {}".format('\n       . '.join(bigrams[-N:])))

# 'business':
  . Most correlated words:
       . growth
       . shares
  . Most correlated collacatios:
       . economic growth
       . stock market
# 'entertainment':
  . Most correlated words:
       . actor
       . film
  . Most correlated collacatios:
       . film festival
       . box office
# 'politics':
  . Most correlated words:
       . election
       . labour
  . Most correlated collacatios:
       . tony blair
       . mr blair
# 'sport':
  . Most correlated words:
       . match
       . cup
  . Most correlated collacatios:
       . australian open
       . champions league
# 'tech':
  . Most correlated words:
       . software
       . users
  . Most correlated collacatios:
       . mobile phones
       . mobile phone


# Classification models

In [224]:
X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size=0.33, random_state = 1000)

Let's try several different classifiers and compare their results

In [228]:
from sklearn import metrics

In [230]:
model_NB = MultinomialNB()
model_NB.fit(X_train, y_train)
pred_NB = model_NB.predict(X_test)
print(metrics.classification_report(y_test, pred_NB))

              precision    recall  f1-score   support

           0       0.95      0.97      0.96       174
           1       1.00      0.93      0.96       127
           2       0.94      0.97      0.95       121
           3       0.99      0.99      0.99       171
           4       0.97      0.97      0.97       142

   micro avg       0.97      0.97      0.97       735
   macro avg       0.97      0.97      0.97       735
weighted avg       0.97      0.97      0.97       735



In [233]:
model_SVC = LinearSVC()
model_SVC.fit(X_train, y_train)
pred_SVC = model_SVC.predict(X_test)
print(metrics.classification_report(y_test, pred_SVC))

              precision    recall  f1-score   support

           0       0.96      0.95      0.96       174
           1       0.99      0.94      0.97       127
           2       0.94      0.98      0.96       121
           3       1.00      1.00      1.00       171
           4       0.95      0.98      0.96       142

   micro avg       0.97      0.97      0.97       735
   macro avg       0.97      0.97      0.97       735
weighted avg       0.97      0.97      0.97       735



In [234]:
model_LR = LogisticRegression()
model_LR.fit(X_train, y_train)
pred_LR = model_LR.predict(X_test)
print(metrics.classification_report(y_test, pred_LR))

              precision    recall  f1-score   support

           0       0.97      0.97      0.97       174
           1       1.00      0.94      0.97       127
           2       0.94      0.97      0.96       121
           3       0.99      1.00      1.00       171
           4       0.97      0.98      0.97       142

   micro avg       0.97      0.97      0.97       735
   macro avg       0.97      0.97      0.97       735
weighted avg       0.97      0.97      0.97       735



In [238]:
model_RF = RandomForestClassifier()
model_RF.fit(X_train, y_train)
pred_RF = model_RF.predict(X_test)
print(metrics.classification_report(y_test, pred_RF))

              precision    recall  f1-score   support

           0       0.89      0.93      0.91       174
           1       0.93      0.88      0.90       127
           2       0.89      0.88      0.89       121
           3       0.95      0.97      0.96       171
           4       0.91      0.89      0.90       142

   micro avg       0.91      0.91      0.91       735
   macro avg       0.91      0.91      0.91       735
weighted avg       0.91      0.91      0.91       735



In [297]:
from sklearn.tree import DecisionTreeClassifier
model_DT = DecisionTreeClassifier()
model_DT.fit(X_train, y_train)
pred_DT = model_DT.predict(X_test)
print(metrics.classification_report(y_test, pred_DT))

              precision    recall  f1-score   support

           0       0.77      0.78      0.77       174
           1       0.84      0.80      0.82       127
           2       0.79      0.83      0.81       121
           3       0.90      0.94      0.92       171
           4       0.85      0.78      0.81       142

   micro avg       0.83      0.83      0.83       735
   macro avg       0.83      0.83      0.83       735
weighted avg       0.83      0.83      0.83       735



In [300]:
from sklearn.svm import SVC
model_SV = SVC()
model_SV.fit(X_train, y_train)
pred_SV = model_SV.predict(X_test)
print(metrics.classification_report(y_test, pred_SV))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00       174
           1       0.00      0.00      0.00       127
           2       0.00      0.00      0.00       121
           3       0.23      1.00      0.38       171
           4       0.00      0.00      0.00       142

   micro avg       0.23      0.23      0.23       735
   macro avg       0.05      0.20      0.08       735
weighted avg       0.05      0.23      0.09       735



C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [299]:
from sklearn.neural_network import MLPClassifier
model_mlp = MLPClassifier()
model_mlp.fit(X_train, y_train)
pred_mlp = model_mlp.predict(X_test)
print(metrics.classification_report(y_test, pred_mlp))

              precision    recall  f1-score   support

           0       0.97      0.97      0.97       174
           1       0.98      0.94      0.96       127
           2       0.95      0.98      0.96       121
           3       1.00      1.00      1.00       171
           4       0.97      0.98      0.97       142

   micro avg       0.98      0.98      0.98       735
   macro avg       0.97      0.97      0.97       735
weighted avg       0.98      0.98      0.98       735



# Model tuning with cross-validation

After testing several models, I took the best four for futher tuning to get even better results: Naive Buyes, Linear Supported Vector Classifier, Multi-Layer Perception and Logistic Regression.

In [270]:
from sklearn.model_selection import GridSearchCV

In [286]:

#creation of the parameter grid for Logistic regression
param_grid_LR = {'penalty': ['l1', 'l2'],'C': [0.0001, 0.001, 0.01, 1, 100]}

# looking for best model
tunning_LR = GridSearchCV(model_LR, param_grid_LR, scoring = 'f1_weighted', cv = 5)
tunning_LR.fit(X_train, y_train)

# checking what is the best model
tunning_LR.best_estimator_
print("Parameters:" ,tunning_LR.best_params_)
print("Score:",tunning_LR.best_score_)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\c

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\c

Parameters: {'C': 100, 'penalty': 'l2'}
Score: 0.9818740736724195


In [287]:
#creation of the parameter grid for Linear SVC
param_grid_SVC = { 'multi_class':['ovr','crammer_singer'],'C': [0.0001, 0.001, 0.01, 1, 100]}

# looking for best model
tunning_SVC = GridSearchCV(model_SVC, param_grid_SVC, scoring = 'f1_weighted', cv = 5)
tunning_SVC.fit(X_train, y_train)

# checking what is the best model
tunning_SVC.best_estimator_
print("Parameters:" ,tunning_SVC.best_params_)
print("Score:",tunning_SVC.best_score_)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\c

Parameters: {'C': 1, 'multi_class': 'ovr'}
Score: 0.9811984740808827


In [289]:
#creation of the parameter grid for Naive Bayes
param_grid_NB = {'alpha' : [0.0001, 0.001, 0.01, 0.1, 1.0]}

# looking for best model
tunning_NB = GridSearchCV(model_NB, param_grid_NB, scoring = 'f1_weighted', cv = 5)
tunning_NB.fit(X_train, y_train)

# checking what is the best model
tunning_NB.best_estimator_
print("Parameters:" ,tunning_NB.best_params_)
print("Score:",tunning_NB.best_score_)

Parameters: {'alpha': 0.01}
Score: 0.9798746316043947


In [ ]:
param_grid_NB = {'alpha' : [0.0001, 0.001, 0.01, 0.1, 1.0],
                'activation':['logistic', 'tahn', 'relu'] ,
                'solver': ['adam', 'sgd','lfbgs'],
                'hidden_layer_sizes': [30, 100]}

# looking for best model
tunning_NB = GridSearchCV(model_NB, param_grid_NB, scoring = 'f1_weighted', cv = 5)
tunning_NB.fit(X_train, y_train)

# checking what is the best model
tunning_NB.best_estimator_
print("Parameters:" ,tunning_NB.best_params_)
print("Score:",tunning_NB.best_score_)

# Final comments

Here I used tdidf approach, instead of 'bag of words' to define the most 'typical' words and phrases for each class and then applied different classifiers to find the best one. Most of applied classifiers performed quite well (with F1-score 0.96+), performing a little bit worse on 'smaller' classes, that might be an issue of some small sample imbalance.
After chosing and tuning several models with cross-validation, I reach the result of F1-score = 0.98+, the best models turned to be Multi-Layer Perception with default settings, LogRegression with C=100 and penalty=l2 and Linear SVC with C=1 and multi_calss=ovr. Possibly, the cross-validation with more than k=5 will show even better results. Nevertheless, all obtained classifier perform pretty good.
The choice of metric is based on the fact that 'F1-score' is rather balanced one compared to precision and recall metrics, which, as we can see, vary from one class to another.

*Note1:the model tuning is self-plagiarism from the first task and possible coincidences with groupmates are really accidental.*

*Note2: after downloading to github I noticed that some errors during running all the file may occure due to some problems with the last libraries updates (there are o lot of similar cases globally with numoy, pandas, seaborn, matplotlib), but it used to run ok earlier. Unfortunately, it can't be fixed by my forces.*